### PR TITLE
SNOW-665404: Improve async job related APIs' annotations

### LINKED ISSUE #505: SNOW-665294: Pandas DataFrame types not working in VS Code
It looks like the recent change of return types here: 

```
 def to_pandas(
        self,
        *,
        statement_params: Optional[Dict[str, str]] = None,
        block: bool = True,
        **kwargs: Dict[str, Any],
    ) -> Union["pandas.DataFrame", AsyncJob]:
```

https://github.com/snowflakedb/snowpark-python/blob/96e96a0fb431a32278d8a9463cd4ba0f1c28b91d/src/snowflake/snowpark/dataframe.py#L608 

causes some issues with VS Code autocomplete which sees this as returning (Any | AsyncJob) which turned my code into a lot of red squiggly lines.

I imagine the double quote string literal of type here is the issue, and previously I expect it was assuming it was type Any so wouldn't auto-complete, but wouldn't red squiggle. The code actually runs fine, but editors are confused by the type being returned.

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #505 

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Now you can get autocomplete after to_pandas()! Also, using @overload can improve the API doc readability:
image

### CODE REVIEW

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -487,10 +488,22 @@ def __init__(
     def stat(self) -> DataFrameStatFunctions:
         return self._stat
 
+    @overload
+    def collect(
+        self, *, statement_params: Optional[Dict[str, str]] = None, block: bool = True
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:493]: I'm curious, given that the default value of `block` is different though the function name is the same, does this mean `block` effectively becomes a required argument?

**Review by sfc-gh-sfan — APPROVED** [review]

**Comment by sfc-gh-yixie** [comment]
Did you try with jupyter notebook and VS code?

**Comment by sfc-gh-jdu** [comment]
> Did you try with jupyter notebook and VS code?

Yes

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -611,6 +636,29 @@ def to_local_iterator(
     def __copy__(self) -> "DataFrame":
         return DataFrame(self._session, copy.copy(self._plan))
 
+    if installed_pandas:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/dataframe.py:639]: What will happen to the IDE/notebook when pandas is not installed?

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/dataframe.py:639]: It's the same as before. Again @overload is for docstring and type checking tools so this PR doesn't have any functional change. When pandas is not installed, only this definition will be skipped, and users are not able to get autocomplete for pandas methods.

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**